Add method to convert internal to stable constructs

celinval opened the pull request:
This is an alternative implementation to https://github.com/rust-lang/rust/pull/116999. I believe we can still improve the logic a bit here, but I wanted to see which direction we should go first.

In this implementation, the API is simpler and we keep Tables somewhat private. The definition is still public though, since we have to expose the Stable trait. However, there's a cost of keeping another thread-local and using `Rc`, but I'm hoping it will be a small cost.

r? @oli-obk
r? @spastorino

celinval committed:
Add method to convert internal to stable constructs

bors commented:
:umbrella: The latest upstream changes (presumably #117013) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

celinval committed:
Add `internal()` method counterpart to `stable()`

Note: We do not expect to provide internalizing methods for all
StableMIR constructs. They exist only to help migrating efforts to allow
users to mix StableMIR and internal constructs.

celinval committed:
Remove unsafe and `Rc`

oli-obk commented on compiler/rustc_smir/src/rustc_internal/mod.rs:
```diff
@@ -125,18 +136,41 @@ pub fn crate_num(item: &stable_mir::Crate) -> CrateNum {
+
+pub(crate) fn init<'tcx>(tables: &TablesWrapper<'tcx>, f: impl FnOnce()) {
+    assert!(!TLV.is_set());
+    let ptr: *const () = &tables as *const &_ as _;
```
This takes a reference to `tables` which is already a reference

celinval replied:
Indeed. I was just replicating the same logic as our current TLV implementation:
https://github.com/rust-lang/rust/blob/41aa06ecf963cf3529146159f761d6e5b5f9cb48/compiler/stable_mir/src/lib.rs#L249

celinval committed:
Add test and remove double ref

oli-obk commented:
@bors r+ rollup

bors commented:
:pushpin: Commit ae86f59cc9dd7d4be3c47d9ecbc3b98ca1f514a8 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk ae86f59cc9dd7d4be3c47d9ecbc3b98ca1f514a8 -->
<!-- homu: {"type":"Approved","sha":"ae86f59cc9dd7d4be3c47d9ecbc3b98ca1f514a8","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
